### ISSUE TITLE
ENH tckgen: parameters via config file

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#1868]
Thought arose while dealing with #1833.

While it's difficult to devise appropriate default tracking parameters that will work across all possible scenarios, it may be the case that some users may consistently use some set of parameters that work for the type of data they are working on at the time; in this case, it would be convenient to not have to manually type out these parameters upon every use, or to modify the macros in the code.

Instead, config file entries could be defined that would override the built-in default tracking parameters/

**Comment by jdtournier** [comment]
I like the general idea, but I'm not sure about requiring users to modify their main config file. Users are not unlikely to forget that their defaults differ - I've been guilty of this in the past. I wouldn't be surprised to find we end up running round in circles trying to debug user issues only to discover they're simply not using our defaults...

What might work is to allow these parameters to be passed through a separate, explicitly specified config file via an environment variable. We could use the same format and backend as for the regular config file(s). Note that this approach would still allow these parameters to reside in the main config, so it doesn't necessarily _prevent_ the issue I'm worried about above. But it might make it (slightly?) less likely. And it would allow users to switch parameters easily between configs (e.g. for different studies). 

We could also extend that to all apps generally - I note that we already have the `MRTRIX_CONFIGFILE` environment variable, which allows users to override the system-wide config file. It'll still look for the user config file in their home folder, but it'll parse this one first _instead_ of the system config (this was added in #1104). We could add another environment variable to allow an additional config files to be parsed _after_ all the standard ones, allowing the user to override any parameters that might have been set in their system-wide or individual config files - but more importantly to preserve all those parameters if they're left untouched in this additional config. All parameters would be as normally used unless explicitly specified in that additional config. To do this, we might have to rename that `MRTRIX_CONFIGFILE` to e.g. `MRTRIX_SYSCONFIG`, and use `MRTRIX_CONFIGFILE` for the additional one to be parsed last. Or shorten it to `MRTRIX_CONFIG`, which might be a good idea to avoid conflict with older versions of _MRtrix3_...

Also: I realise this is pretty tangential to the original point of this issue... But I'm all for it, I just think we would need to be a bit careful with how it might be implemented.

**Comment by Lestropie** [comment]
> Users are not unlikely to forget that their defaults differ

Would this be adequately addressed by a `CONSOLE()` call stating that default tracking parameters have been changed based on the config file (preferably only in cases where at least one such modified default is consequential)?

**Comment by jdtournier** [comment]
> Would this be adequately addressed by a CONSOLE() call stating that default tracking parameters have been changed based on the config file 

Yes and no. I can see this might cover us more easily if we come to debug issues arising from this, but I reckon if I was to do this, I'd quickly get irritated with this. Might be better as an INFO level output - so it'll at least show up in the output if we request they produce `-info` or `-debug` output... What do you think?

**Comment by Lestropie** [comment]
I've no issue with lowering the verbosity; `-info` is a common recommendation for `tckgen` assessment because it gives the termination & rejection stats.

**Comment by Lestropie** [comment]
Could do this multi-layered:

- *MRtrix3* config file could have an entry that contains the *path* to another config file containing the `tckgen`-specific fields;

- Environment variable could override this at runtime.
